### PR TITLE
Introduce IRemoteContext

### PR DESCRIPTION
**Pull request by ilyachur**
### Details:
 - Changed plugin API
 - Added Remote tensor implementation for template plugin
 - Replaced PR https://github.com/openvinotoolkit/openvino/pull/16178

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/inference/src/dev/converter_utils.cpp [hunk 1]
```diff
@@ -789,3 +791,99 @@ std::shared_ptr<::ov::IAsyncInferRequest> ov::legacy_convert::convert_infer_requ
...
+        return m_context->get_device_name();
+    }
+
+    InferenceEngine::RemoteBlob::Ptr CreateBlob(const InferenceEngine::TensorDesc& tensorDesc,
+                                                const InferenceEngine::ParamMap& params = {}) override {
+        OPENVINO_NOT_IMPLEMENTED;
+        // return ov::tensor_to_blob(
+        //     m_context->create_tensor(InferenceEngine::details::convertPrecision(tensorDesc.getPrecision()),
+        //                              tensorDesc.getBlockingDesc().getBlockDims(),
+        //                              params));
```
Comment (ilya-lavrenov) [inline src/inference/src/dev/converter_utils.cpp:818]: should we uncomment this code?

File: src/plugins/template/src/remote_context.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        break;
+    case ov::element::u16:
+        tensor = std::make_shared<VectorTensorImpl<ov::element_type_traits<ov::element::u16>::value_type>>(type, shape);
+        break;
+    case ov::element::u32:
+        tensor = std::make_shared<VectorTensorImpl<ov::element_type_traits<ov::element::u32>::value_type>>(type, shape);
+        break;
+    case ov::element::u64:
+        tensor = std::make_shared<VectorTensorImpl<ov::element_type_traits<ov::element::u64>::value_type>>(type, shape);
+        break;
```
Comment (ilya-lavrenov) [inline src/plugins/template/src/remote_context.cpp:179]: why do we need template code? can we have simpler example / implementation ?

File: src/plugins/template/include/template/remote_tensor.hpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+     * @brief Checks that type defined runtime parameters are presented in remote object
+     * @param tensor a tensor to check
+     */
+    static void type_check(const Tensor& tensor) {
+        RemoteTensor::type_check(tensor, {{ov::device::id.name(), {"TEMPLATE"}}});
+    }
+
+    /**
+     * @brief Returns the underlying vector
+     * @return vector if T is compatible with element type
```
Comment (ilya-lavrenov) [inline src/plugins/template/include/template/remote_tensor.hpp:29]: is this documentation string right?

File: src/plugins/template/src/compiled_model.cpp [hunk 4]
```diff
@@ -71,8 +71,8 @@ std::shared_ptr<ov::ISyncInferRequest> ov::template_plugin::CompiledModel::creat
 void ov::template_plugin::CompiledModel::set_property(const ov::AnyMap& properties) {
     OPENVINO_NOT_IMPLEMENTED;
 }
-ov::RemoteContext ov::template_plugin::CompiledModel::get_context() const {
-    OPENVINO_NOT_IMPLEMENTED;
+std::shared_ptr<ov::IRemoteContext> ov::template_plugin::CompiledModel::get_context() const {
+    return get_template_plugin()->get_default_context({});
```
Comment (ilya-lavrenov) [inline src/plugins/template/src/compiled_model.cpp:75]: if model was created with remote context, it should store it internally.

File: src/plugins/template/src/remote_context.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    return impl->get_data_ptr();
+}
+
+void* ov::template_plugin::VectorTensor::get_data_ptr() {
+    auto impl = std::dynamic_pointer_cast<VectorImpl>(_impl);
+    OPENVINO_ASSERT(impl);
+    return impl->get_data_ptr();
+}
+
+ov::template_plugin::RemoteContext::RemoteContext() : m_name("TEMPLATE") {}
```
Comment (ilya-lavrenov) [inline src/plugins/template/src/remote_context.cpp:124]: plugin library cannot include implementation of it's public API

**Review by ilya-lavrenov — APPROVED** [review]

File: src/inference/src/dev/icompiled_model.cpp [hunk 6]
```diff
@@ -4,6 +4,9 @@
 
 #include "openvino/runtime/icompiled_model.hpp"
 
+#include <openvino/core/any.hpp>
```
Comment (ilyachur) [inline src/inference/src/dev/icompiled_model.cpp:7]: TODO: remove this include

File: src/inference/src/dev/isync_infer_request.cpp [hunk 7]
```diff
@@ -13,6 +13,7 @@
 #include "openvino/op/util/op_types.hpp"
 #include "openvino/runtime/icompiled_model.hpp"
 #include "openvino/runtime/iinfer_request.hpp"
+#include "openvino/runtime/iremote_context.hpp"
 #include "openvino/runtime/remote_context.hpp"
```
Comment (ilyachur) [inline src/inference/src/dev/isync_infer_request.cpp:17]: Looks like I can remove this header

File: src/plugins/template/include/template/remote_tensor.hpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        OPENVINO_ASSERT(params.count("vector_data"), "Cannot get data. Tensor is incorrect!");
+        try {
+            auto& vec = params.at("vector_data").as<std::vector<T>>();
+            return vec;
+        } catch (const std::bad_cast&) {
+            OPENVINO_THROW("Cannot get data. Vector type is incorrect!");
+        }
+    }
+
+    const void* get_data() const {
```
Comment (ilyachur) [inline src/plugins/template/include/template/remote_tensor.hpp:64]: TODO: add doxygen

**Pull request merged**